### DISCUSSION TITLE
Circular references cannot be expressed in JSON.

### DISCUSSION THREAD

**Discussion by Zorig** [vercel/next.js#10992] [Help]
Hey guys, I faced some issues when using `getServerSideProps` on 9.3. I fetched some data using apollo graphql and pass it to the page. but i got following error.
```
Error serializing `.mixes.items[2].user` returned from `getServerSideProps` in "/". Reason: Circular references cannot be expressed in JSON.
Error: Error serializing `.mixes.items[2].user` returned from `getServerSideProps` in "/".
Reason: Circular references cannot be expressed in JSON.
    at visit (/Users/zorig/xyyp/play/node_modules/next/dist/lib/is-serializable-props.js:1:668)
    at isSerializable (/Users/zorig/xyyp/play/node_modules/next/dist/lib/is-serializable-props.js:7:256)
    at /Users/zorig/xyyp/play/node_modules/next/dist/lib/is-serializable-props.js:7:456
```

my fetched data looks like the following:
```
 mixes: {
    items:[
        {
          __typename: "Mix",
          id: "1",
          title: "Хип Хоп хит дуунууд",
          thumb: "https://localhost:8000/media/mix/hhrnb.jpg",
          user: {
            __typename: "User",
            id: "1",
            name: "User1"
          }
        },
      {
          __typename: "Mix",
          id: "2",
          title: "Dance EDM",
          thumb: "https://localhost:8000/media/dance_mix.jpg",
          user: {
            __typename: "User",
            id: "2",
            name: "User2"
          }
        },
        {
          __typename: "Mix",
          id: "4",
          title: "Acoustic Vibes",
          thumb: "https://localhost:8000/media/Acoustic_vibes_mix.jpg",
          user: {
            __typename: "User",
            id: "2",
            name: "User2"
          }
        },
    ]
}
```

and i tried to run inspection and run `is-serializable-props` on my fetched data.

```javascript
import { isSerializableProps } from "next/dist/lib/is-serializable-props";
    isSerializableProps("adasd", "asdasd", {
      props: {
         mixes
      }
```
which throws the exact same error. But when i hardcoded data inside props, it does not throw an error

I even checked by data by  `isPlainObject` from lodash and it shows true.

**Comment by Timer** [comment]
Please upgrade to `next@canary` where we fixed a false positive. Sorry about that!

**Comment by giovannipds** [comment]
Hi. I just went through this on `getStaticProps` with latest `next@canary` (`9.5.3-canary.25`). I've solved it with `decycle`/`retrocycle` functions after reading about them on [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Cyclic_object_value).

```js
import { decycle, retrocycle } from "cycle";

export default function Page({ entry }) {
  const fineEntry = retrocycle(entry);
  // ...
}

export async function getStaticProps() {
  const entry = decycle(cycledEntry);
  return {
    props: {
      entry,
    },
  };
}
```

**Comment by AlexandraKlein** [comment]
I'm having this issue when fetching data from Contentful on posts that have links to each other. 

Re: https://github.com/vercel/next.js/blob/86160a5190c50ea315c7ba91d77dfb51c42bc65f/packages/next/lib/is-serializable-props.ts#L41

Stringifying using `safe-json-stringify` and reparsing fixes, but I'm hoping there is a more performant solution.

```javascript
import safeJsonStringify from 'safe-json-stringify';
...

export async function getServerSideProps({ params, preview = false }) {
  const rawData = await getSlugPageProps(params.slug, 'product-page', preview);
  const stringifiedData = safeJsonStringify(rawData);
  const data = JSON.parse(stringifiedData);

  return {
    props: {
      data
    }
  };
}
```

**Reply by giovannipds** [reply]
Yep, mine was to solve integration with Contentful too.

**Reply by AlexandraKlein** [reply]
Was hoping there was a more elegant solution for this. Let me know if you come up with one? And I'll do the same. 😃

**Reply by giovannipds** [reply]
Of course, I'll stay with `decycle` / `retrocycle` while no official solution is proposed.

**Reply by coryetzkorn** [reply]
Same problem here, using Contentful

**Reply by ChazUK** [reply]
Thanks for making this post! I ended up using https://github.com/WebReflection/flatted as it also reduced the size of the string.

**Comment by AlexandraKlein** [comment]
I'm finding that stringifying data is not an optimal solution for me. I have posts that relate to posts, which have their own related posts, creating an infinitely deep object. This seems like a Contentful problem to me...


```
items: {
    0: {
      fields: {
        title: 'article 1',
        relatedPosts: {
          0: {
            fields: {
              title: 'article 2',
              relatedPosts: {
                0: {
                  fields: {
                    title: 'article 3',
                    relatedPosts: {
                      0: {
                        fields: {
                          title: 'article 4',
                          relatedPosts: [....goes on very deep....]
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
```

**Reply by giovannipds** [reply]
Did you try `decycle`/`retrocycle`?

**Reply by AlexandraKlein** [reply]
When I use `decycle/retrocycle`, it returns the circular reference, but as far as I can tell, this would still cause an incredibly deep object when there are many posts. If each post has 3 related posts, this causes an incredibly deep object, whether they are circular references or not.

**Reply by giovannipds** [reply]
Hi @AlexandraKlein. Thinking about this issue, I would say that you're probably thinking about it in the wrong way. I can't imagine a use case you'll really need all this depth of object. Probably the best way to solve your case is to limit that depth on the request to Contenful, just that. There's a good explanation about it [here](https://www.contentful.com/developers/docs/javascript/tutorials/using-js-cda-sdk/#retrieve-linked-entries), make sure to check it out and re-think about your case.

**Reply by AlexandraKlein** [reply]
@giovannipds  The include property does not respect linked content types. If I `include: 2`, the linked `relatedPosts` still returns a very deep object, much deeper than 2 levels.

If I use `include: 1`, issue is resolved, but I can't access other content types at a depth of 2 that I need to access.

I opened an issue here: https://github.com/contentful/contentful.js/issues/439

**Reply by giovannipds** [reply]
Again, exactly as they said there, if you have something circular, it'll go on and on, of course. Why wouldn't use `include: 1` and make additional requests if you really need more of it? It'd be great if you share exactly your use case. I still think you're trying to solve it with just one request when you could possibly limit + make more requests if needed.

**Reply by giovannipds** [reply]
Think about related posts anyway. They are RELATED, they are NOT the main content. It shouldn't be loaded the way you're thinking there, they're probably `aside`. That's my point...

**Reply by AlexandraKlein** [reply]
That is exactly how I'm solving it right now:  Separating and implementing additional requests. I was hoping for a more elegant solution. Thanks for all your engagement and help, @giovannipds!

**Reply by giovannipds** [reply]
No trouble. I'm sorry if anytime I may have sounded rude but, that was not the intention. I really don't think it'd be important to have a better solution than what you're already doing - considering the case posts / related posts. Congrats on that. Keep it going. 🤘

**Comment by boJackEden** [comment]
Has this been fixed with the release of Next 10?

**Comment by eduoliveira85** [comment]
I've been building a Next.js site with Contentful intregation for a couple months now and am using this solution [here](https://github.com/vercel/next.js/discussions/10992#discussioncomment-103826), suggested by @AlexandraKlein. Everything was working fine until today, when I went back to my blog front page to find a similar error to the one which brought me here at first, couple months ago. Is anyone experiencing this same problem?

Console error:

> Server Error
> Error: Error serializing `.posts` returned from `getStaticProps` in "/blog".
> Reason: `undefined` cannot be serialized as JSON. Please use `null` or omit this value.
> 
> This error happened while generating the page. Any console logs will be displayed in the terminal window.

**Reply by AlexandraKlein** [reply]
Have you tried parsing your JSON in the component? That is, moving `const data = JSON.parse(stringifiedData);` from `getStaticProps` into the body of your component and return just the stringified JSON from `getStaticProps`?

**Reply by eduoliveira85** [reply]
HI @AlexandraKlein. I've just tried what you suggested and the error changed to the first one I saw before finding the solution you proposed.

> Server Error
> Error: Error serializing `.postsData[8].fields.excerpt.content[0].content[1].data.target` returned from `getStaticProps` in "/blog".
> Reason: Circular references cannot be expressed in JSON (references: `.postsData[8]`).
> 
> This error happened while generating the page. Any console logs will be displayed in the terminal window.

**Reply by AlexandraKlein** [reply]
You tried this?

```javascript
function someComponent({ data }) {
  const posts = JSON.parse(data);

  return (
    <div>
      {posts.map(post => (
        <h1>{post.title}</h1>
      ))}
    </div>
  );
}

export async function getStaticProps() {
  const rawData = await yourFunctionToGetData();
  const data = stringify(rawData);

  return {
    props: {
      data
    }
  };
}

```

**Reply by AlexandraKlein** [reply]
If that doesn't work, you probably have to replace all your `undefined`s in your response with `null`.

**Reply by eduoliveira85** [reply]
It didn't work for me either way. I'm ditching contentful for this project. Thanks you for your kind reply!

**Reply by Sergioamjr** [reply]
@eduoliveira85 just for you know I've been facing the same problem with Next and Contentful this morning on the `getStaticProps` method.
```
Server Error
Error: Error serializing `XXX` returned from `getStaticProps` in "/join-us".
Reason: Circular references cannot be expressed in JSON (references: `.page.allMembers`).
```

I figured out this problem is caused when you try to access a page with a link for itself and return the fields in more than one level.

For instance, I have this query.

```js
const data = await client.getEntries({
  content_type: 'pageJoinUs',
  include: 2,
});
```

If I change the `include` to 1, I won't get this error. Neither the fields that are references to other content type on Contentful. I'll receive their sysid and have to fetch their fields in other request.

It's also possible to change the href of the link that refers to the same page. It'll work even with `include: 2`.

I hope this help someone.

**Comment by Webcoda** [comment]
If you want to keep it simple (to remove the circular dependency and you only need A - B relationship instead of A - B - A), you can use `fast-safe-stringify`. Writing this when using Kentico Kontent with Next.js
